hail/src/main/scala/is/hail/expr/ir/agg/CallStatsAggregator.scala

```diff
@@ -129,37 +129,33 @@ class CallStatsAggregator(t: PCall) extends StagedAggregator {
 
   def seqOp(state: State, seq: Array[EmitCode], dummy: Boolean): Code[Unit] = {
     val Array(call) = seq
-    val hom = state.cb.genFieldThisRef[Boolean]()
-    val lastAllele = state.cb.genFieldThisRef[Int]()
-    val i = state.cb.genFieldThisRef[Int]()
-
-    def checkSize(a: Code[Int]): Code[Unit] =
-      Code.memoize(a, "callstatsagg_check_size_a") { a =>
-        (a > state.nAlleles).orEmpty(
-          Code._fatal[Unit](const("found allele outside of expected range [0, ")
-            .concat(state.nAlleles.toS).concat("]: ").concat(a.toS)))
+    assert(t == call.pv.pt)
+
+    val mb = state.cb.getOrGenEmitMethod("callstatsagg_seqop", t, Array(typeToTypeInfo(t)), UnitInfo) { mb =>
+      mb.emitWithBuilder[Unit] { cb =>
+        val hom = cb.memoize[Boolean](const(true), "hom")
+        val lastAllele = cb.memoize[Int](const(-1), "lastAllele")
+        val i = cb.memoize[Int](const(0), "i")
+        val call = PCallValue(t, mb.getArg(1)(typeToTypeInfo(t)))
+
+        call.forEachAllele(cb) { (cb: EmitCodeBuilder, allele: Value[Int]) =>
```

Comment:
Oh, this is superb.  Nicely done.

One small comment: I think codegen emitters (the second argument to forEachAllele here) shouldn't take the code builder, because in almost all cases, they will be defined in an environment that has a code builder in scope.  I started writing everything in this style and there were code builder bindings everywhere that all referred to the same object.  Totally unnecessary.

Reply:
Ok. So it just becomes the contract that the closure needs to put code onto the same CodeBuilder as is passed in. Easy enough.